### PR TITLE
feat(knowledge): implement transitive trust rules (PR14/21)

### PR DESCRIPTION
**Pull request by H31MDALLR**
## Summary

Implements N1 §2.10.2 transitive trust rules to prevent trust escalation attacks in knowledge packs.

**This is PR14/21 in the ETL & Trust enhancement sequence** - one of 4 critical PRs for OSS v1.0.

## Context

Spec enhancement PR #84 added comprehensive trust model requirements. This PR implements the core transitive trust rules that prevent malicious packs from escalating privileges through dependency chains.

### Files Created (588 lines)

1. **`components/knowledge/src/ai/miniforge/knowledge/trust.clj`** (303 lines)
   - Trust level ordering: `:tainted` < `:untrusted` < `:trusted`
   - Authority channels: `:authority/instruction` vs `:authority/data`
   - Pack reference creation and validation
   - Circular dependency detection (DFS)
   - Tainted content isolation verification

2. **`components/knowledge/test/ai/miniforge/knowledge/trust_test.clj`** (285 lines)
   - 59 assertions testing all 4 transitive trust rules
   - Edge case coverage (circular deps, missing deps, tainted isolation)
   - Trust level inheritance validation

### Files Modified (252 lines)

3. **`components/knowledge/src/ai/miniforge/knowledge/interface.clj`** (+62 lines)
   - Exported trust validation API: `make-pack-ref`, `validate-transitive-trust`, `compute-inherited-trust-level`

4. **`components/policy-pack/src/ai/miniforge/policy_pack/loader.clj`** (+170 lines)
   - Layer 4: Trust validation integration
   - `validate-pack-trust` with dependency resolution
   - Trust-enforced pack loading

5. **`components/policy-pack/src/ai/miniforge/policy_pack/schema.clj`** (+24 lines)
   - `TrustLevel` and `AuthorityChannel` schemas
   - Enhanced `PackManifest` with trust fields

## The 4 Transitive Trust Rules

1. ✅ **Instruction authority is not transitive**
   - Trusted pack with instruction authority cannot grant instruction authority to untrusted dependencies

2. ✅ **Trust level inheritance**
   - Combined content takes lowest trust level (tainted < untrusted < trusted)

3. ✅ **Cross-trust references**
   - Validates complete dependency graph before execution
   - Detects circular dependencies and missing deps

4. ✅ **Tainted isolation**
   - Tainted content cannot be used for instruction authority (checked transitively)

## Test Results

- ✅ All 59 trust-specific assertions passing
- ✅ All existing tests continue to pass
- ✅ Pre-commit validation passed

## Acceptance Criteria

- ✅ All 4 transitive trust rules enforced
- ✅ Trusted pack cannot transitively grant instruction authority to untrusted pack
- ✅ Trust level inheritance uses lowest trust in chain
- ✅ Tainted content isolated from instruction authority
- ✅ Tests verify all rules

## Dependencies

This PR is independent and can merge in parallel with:
- PR15: Pack dependency validation
- PR16: ETL lifecycle events
- PR17: Promotion justification

## Spec Conformance

- ✅ N1 §2.10.2 (Knowledge Trust and Authority) - FULL CONFORMANCE

## Impact

- **Lines changed:** +840
- **Files changed:** 5
- **Breaking changes:** None (additive only)
- **Blocks:** OSS v1.0 release

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: components/knowledge/src/ai/miniforge/knowledge/trust.clj [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+   Returns:
+   - {:valid? true :graph {...}} if valid
+   - {:valid? false :error \"...\"} if invalid (circular deps, etc.)"
+  [pack-graph]
+  (let [visited (atom #{})
+        visiting (atom #{})]
+
+    ;; Detect circular dependencies via depth-first search
+    (letfn [(visit [pack-id path]
```
Comment (H31MDALLR) [inline components/knowledge/src/ai/miniforge/knowledge/trust.clj]: If we do not constrain the size of the graph then loop/recur will avoid stack overflow.

File: components/knowledge/src/ai/miniforge/knowledge/trust.clj [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+   track and validate the transitive trust graph.
+
+   Arguments:
+   - pack-graph - Map of pack-id -> pack-ref
+
+   Returns:
+   - {:valid? true :graph {...}} if valid
+   - {:valid? false :error \"...\"} if invalid (circular deps, etc.)"
+  [pack-graph]
+  (let [visited (atom #{})
```
Comment (H31MDALLR) [inline components/knowledge/src/ai/miniforge/knowledge/trust.clj]: Atoms are not needed for graph search. This seems arbitrary convenience. Idiomatic clojure would use a vector for depth first and a persistent queue for bfs. I would change this to avoid the mutation unless you really need it.

File: components/knowledge/src/ai/miniforge/knowledge/trust.clj [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+   Arguments:
+   - pack-id     - Pack to validate
+   - pack-graph  - Map of pack-id -> pack-ref
+
+   Returns:
+   - {:valid? true} if no tainted content in instruction chain
+   - {:valid? false :error \"...\" :tainted-path [...]} if tainted found"
+  [pack-id pack-graph]
+  (let [visited (atom #{})]
```
Comment (H31MDALLR) [inline components/knowledge/src/ai/miniforge/knowledge/trust.clj]: Again with the atom. Reference impl:

`clojure
(defn neighbors [graph vertex] (-> graph vertex (subvec 1)))
(defn visited? [key coll] (some #{key} coll))

(defn graph-search [graph start found? storage]
  (loop [stack (conj storage start)
         visited []]
    (if (empty? stack)
      nil
      (let [v (peek stack)
            neighbors (-> graph v)
            not-visited (filter (complement #(visited? % visited)) neighbors)
            new-stack (into (pop stack) not-visited)]
        (log/trace ::traverse {:node v :neighbors neighbors :new-stack (seq new-stack)})
        (cond
          (found? v) v
          (visited? v visited) (recur new-stack visited) ;; we've been here before, skip
          :else (recur new-stack (conj visited v)) ;; new visit, add to visited.
          )))))

(defn graph-search-II
  "Walk a graph constructred of Node records and return the output."
  [graph start found? storage]
  (loop [stack (conj storage start)
         visited []]
    (if (empty? stack)
      :not-found
      (let [v (peek stack)
            neighbors (-> graph v)
            new-stack (into (pop stack) neighbors)]
        (log/trace ::traverse {:node v :neighbors neighbors :new-stack (seq new-stack)})
        (cond
          (found? v) v
          (visited? v visited) (recur new-stack visited) ;; we've been here before, skip
          :else (recur new-stack (conj visited v)) ;; new visit, add to visited.
          )))))

(defn dfs
  "Depth first search on a graph, calling fn on each vertex."
[graph vertex f]
(graph-search graph vertex f []))

(defn bfs
  [graph vertex f]
  (graph-search graph vertex f clojure.lang.PersistentQueue/EMPTY))`

Comment (H31MDALLR) [reply components/knowledge/src/ai/miniforge/knowledge/trust.clj]: I feel like we need a DFS implementation here that takes in the ruleset of predicates to test against (or something similar) instead of rewriting DFS each time (and using atoms).

File: components/knowledge/src/ai/miniforge/knowledge/trust.clj (hunk 1 above)
Comment (H31MDALLR) [reply components/knowledge/src/ai/miniforge/knowledge/trust.clj]: ✅ Resolved: Refactored to use loop/recur pattern which avoids stack overflow on large graphs. The new implementation uses immutable data structures (visited and visiting sets) that are passed through the recursion instead of atoms.

File: components/knowledge/src/ai/miniforge/knowledge/trust.clj (hunk 2 above)
Comment (H31MDALLR) [reply components/knowledge/src/ai/miniforge/knowledge/trust.clj]: ✅ Resolved: Refactored validate-cross-trust-references to use loop/recur with immutable data structures. The new implementation uses visited/visiting sets passed through recursion, eliminating all atom usage. This is now idiomatic Clojure and prevents stack overflow on large graphs.

File: components/knowledge/src/ai/miniforge/knowledge/trust.clj (hunk 3 above)
Comment (H31MDALLR) [reply components/knowledge/src/ai/miniforge/knowledge/trust.clj]: ✅ Resolved: Refactored validate-tainted-isolation to use loop/recur pattern similar to your reference implementation. The new code uses immutable visited sets passed through recursion, eliminating all atom usage. Thanks for the reference - it helped guide the idiomatic refactor!

Comment (H31MDALLR) [reply components/knowledge/src/ai/miniforge/knowledge/trust.clj]: ✅ Resolved: Refactored validate-transitive-trust to use functional composition (keep, mapcat, concat) instead of atoms for error collection. The implementation now uses the DFS functions (validate-cross-trust-references and validate-tainted-isolation) which are themselves implemented using loop/recur. All atom usage has been eliminated from the trust validation code.

Comment (H31MDALLR) [reply components/knowledge/src/ai/miniforge/knowledge/trust.clj]: ✅ Updated: Extracted DFS implementations into reusable generic functions:

```clojure
;; Generic DFS to find nodes matching a predicate
(defn ^:private dfs-find [graph start-id get-deps-fn found-fn] ...)

;; Generic DFS to validate graph structure (cycles, missing nodes)
(defn ^:private dfs-validate-graph [graph start-nodes get-deps-fn validate-fn] ...)
```

Both `validate-cross-trust-references` and `validate-tainted-isolation` now use these generic DFS functions with predicates passed in. Also extracted the rule checking logic from `validate-transitive-trust` into separate functions (`collect-authority-errors` and `collect-tainted-errors`) for better readability.

See commit 2af1f4e for full implementation.

Comment (H31MDALLR) [reply components/knowledge/src/ai/miniforge/knowledge/trust.clj]: ✅ Further improved: Extracted canonical `dfs-traverse` function that eliminates ALL nested loops:

```clojure
(defn ^:private dfs-traverse
  "Canonical DFS traversal with pluggable visit function."
  [graph start-nodes get-deps-fn visit-fn]
  ;; Uses reduce instead of nested loop/recur
  ...)
```

Changes:
- Single canonical DFS implementation used by both `dfs-find` and `dfs-validate-graph`
- No more nested loops - replaced with `reduce` for cleaner, more functional code
- Pluggable visit-fn allows different behaviors (find vs validate)
- More idiomatic Clojure

See commit 7dd3f86 for implementation.

Comment (H31MDALLR) [reply components/knowledge/src/ai/miniforge/knowledge/trust.clj]: ✅ Simplified further: Replaced with single canonical `dfs` helper (commit 1964129):

```clojure
(defn ^:private dfs
  "Simple canonical DFS traversal."
  [graph start-ids get-deps-fn on-visit-fn on-cycle-fn on-missing-fn]
  ;; Single DFS implementation with 3 simple hook functions
  ...)
```

Both `dfs-find` and `dfs-validate-graph` now just call this canonical DFS with different hook implementations. Much simpler - the DFS logic is in one place, behaviors are in simple hook functions.

Uses loop/recur (not reduce) for clarity as you suggested.
